---
layout: post
title: "Genetic Bayesian Network (IV)"
date: 2013-11-06 08:01
comments: true
categories: PGN, genetic
---

[UPDATE] I have updated a posteriori the code to include the vectorized version of Factor manipulations, which is incredibly faster and will be explained in a later post.

## example revisited: decoupled BN for cystic fibrosis

{% img center /images/nov13/decoup_cysticBN.png Template for Decoupled Genetic Bayesian Network%}

Let's run our previous example through our decoupled model with a twist. Now instead of two possible alleles (F and f) we are going to have 3: (F, f, n) with the following probabilities to develop CF. Everything else runs similarly as before.

|---|----|-----|-----|-----|----|----|
||||||||
|:--|:--:|:---:|:---:|:---:|:--:|:--:|
|---|----|-----|-----|-----|----|----|
| genotype   |FF  | Ff  | Fn  | ff  | fn | nn |
| prob CF    |80% | 60% | 10% | 50% | 5% | 1% |
|---|----|-----|-----|-----|----|----|
{:.widetable}
<br/>


```python building decoupled network
# family tree and input data
family_tree = {}
family_tree['Ira'] = [None, None]
family_tree['Robin'] = [None, None]
#family_tree['Aaron'] = [None, None]
family_tree['Rene'] = [None, None]
family_tree['James'] = ['Ira', 'Robin']
family_tree['Eva']   = ['Ira', 'Robin']
#family_tree['Sandra'] = ['Aaron', 'Eva']
family_tree['Jason'] = ['James', 'Rene']   # to speed things up
family_tree['Benito'] = ['James', 'Rene']

frequency_of_alleles_in_general_population = [0.1, 0.7, 0.2]
probability_of_trait_based_on_genotype = [0.8, 0.6, 0.1, 0.5, 0.05, 0.01]

TRAIT_PRESENT, TRAIT_ABSENT = [0,1]
F,f,n = [0,1,2]

# building the network
GN = geneticNetwork(family_tree, frequency_of_alleles_in_general_population, probability_of_trait_based_on_genotype)
```

In an initial state, without more information, the probability for Benito developing Cystic Fibrosis is 35.54%.

Now, his grandfather Ira had the ailment, (the trait was present as expressed in phenotype and suffered the disease), and so the chances of Benito being in the same situation raises to 39.01%.

Furthermore, his mother did a genetic test prior to having children and knows that she has a genotype Ff. Now the probability has risen again to 51.47%.

Finally, we do know that aunt Eva developed Cystic Fibrosis. This gives us more information about the probability of James inheriting bad genes from Ira, and so the probability of Benito developing Cystic Fibrosis raises again to 54.10%.

Finally consider that with the unvectorized version of the Factor code I could not run the whole family tree because of memory and time constraints, which forced me to comment out some nodes to run the example in a smaller pedigree tree. Now, commenting out 2 nodes of the family tree (Aaron and Sandra) it still takes quite a few minutes to run it, and the whole CPD has 20 random variables (matrix dimensions) and 204 million cells (different assignments, each one with a probability).

```python conditioning and inferring
# Evidence conditioning
modify_Factor_by_evidence('Ira',  'pheno', TRAIT_PRESENT)
modify_Factor_by_evidence('Rene', 'geno1', F)
modify_Factor_by_evidence('Rene', 'geno2', f)
modify_Factor_by_evidence('Eva',  'pheno', TRAIT_PRESENT)

# lets try first the whole kahuna CPD and compute the prob of developing CF
a = build_joint_cpd()

# marginalizing
target = GN['Benito']['var_pheno']
lista = [x for x in a.variables if x!=target]
for v in lista:
    a = FactorOperations.marginalize(a, v)

print 'probability of Benito showing ailment', 100.*a.values[0], '%'
```

