
<!DOCTYPE html>
<!--[if IEMobile 7 ]><html class="no-js iem7"><![endif]-->
<!--[if lt IE 9]><html class="no-js lte-ie8"><![endif]-->
<!--[if (gt IE 8)|(gt IEMobile 7)|!(IEMobile)|!(IE)]><!--><html class="no-js" lang="en"><!--<![endif]-->
<head>
  <meta charset="utf-8">
  <title>Genetic Bayesian Network (II) - SotoSeattle</title>
  <meta name="author" content="Javier Soto">

  
  <meta name="description" content="[UPDATE] I have updated a posteriori the code to include the vectorized version of Factor manipulations, which is incredibly faster and will be &hellip;">
  

  <!-- http://t.co/dKP3o1e -->
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  
  <link rel="canonical" href="http://sotoseattle.github.io/blog/2013/11/04/GBN2">
  <link href="/favicon.png" rel="icon">
  <link href="/stylesheets/screen.css" media="screen, projection" rel="stylesheet" type="text/css">
  <link href="" rel="alternate" title="SotoSeattle" type="application/atom+xml">
  <script src="/javascripts/modernizr-2.0.js"></script>
  <script src="//ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
  <script>!window.jQuery && document.write(unescape('%3Cscript src="./javascripts/lib/jquery.min.js"%3E%3C/script%3E'))</script>
  <script src="/javascripts/octopress.js" type="text/javascript"></script>
  <!--Fonts from Google"s Web font directory at http://google.com/webfonts -->
<link href="http://fonts.googleapis.com/css?family=PT+Serif:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href="http://fonts.googleapis.com/css?family=PT+Sans:regular,italic,bold,bolditalic" rel="stylesheet" type="text/css">
<link href='http://fonts.googleapis.com/css?family=Open+Sans' rel='stylesheet' type='text/css'>
<link href='http://fonts.googleapis.com/css?family=Fjalla+One' rel='stylesheet' type='text/css'>
<!-- MathJax -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [ ['$','$'], ["$$","$$"] ],
      processEscapes: true
    }
  });
</script>

<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      tex2jax: {
        skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
      }
    });
</script>

<script type="text/x-mathjax-config">
    MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for(i=0; i < all.length; i += 1) {
            all[i].SourceElement().parentNode.className += ' has-jax';
        }
    });
</script>

<script type="text/javascript"
   src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>

  

</head>

<body   class="collapse-sidebar sidebar-footer" >
  <header role="banner">
	<div class="header-title"><a href="/">SotoSeattle</a></div>


	<br><div class="header-subtitle">Software Interloper Developer</div>

</header>
  <nav role="navigation"><ul class="subscription" data-subscription="rss">
<!--   <li><a href="" rel="subscribe-rss" title="subscribe via RSS">RSS</a></li> -->
  
  <li class='icon'>
    <a href="https://github.com/sotoseattle/">
      <img src="/images/logos/github_icon.png" width="36" height="27">
      </img>
    </a>
  </li>
  <li class='icon'>
    <a href="https://www.linkedin.com/in/sotoseattle/">
      <img src="/images/logos/linkedin-256.png"  width="27" height="27">
      </img>
    </a>
  </li>
</ul>
  
<form action="http://google.com/search" method="get">
  <fieldset role="search">
    <input type="hidden" name="q" value="site:sotoseattle.github.io" />
    <input class="search" type="text" name="q" results="0" placeholder="Search"/>
  </fieldset>
</form>
  
<ul class="main-navigation">
<!--   <li><a href="projects/">Projects</a></li> -->
  <li><a href="/">Blog</a></li>
  <li><a href="/blog/archives">Archives</a></li>
  <li><a href="/images/Javier_Soto_resume.pdf">Resume</a></li>
  <li><a href="/about">About</a></li>
</ul>

</nav>
  <div id="main">
    <div id="content">
      <div>
<article class="hentry" role="article">
  
  <header>
  
    
      <h1 class="entry-title">Genetic Bayesian Network (II)</h1>
    
  
    
      <p class="meta">
        








  


<time datetime="2013-11-04T08:01:00-08:00" pubdate data-updated="true">Nov 4<span>th</span>, 2013</time>
        
      </p>
    
  </header>


<div class="entry-content"><p>[UPDATE] I have updated a posteriori the code to include the vectorized version of Factor manipulations, which is incredibly faster and will be explained in a later post.</p>

<h2 id="an-example-cystic-fibrosis">an example: cystic fibrosis</h2>

<p>Let’s create a network for the Cystic Fibrosis. It is expressed through two possible alleles: F and f, where F is dominant. The alleles can take the following genetic forms: FF, Ff, and ff. Given a pedigree like the one bellow lets build a Bayesian network and run some inference.</p>

<p><img class="center" src="/images/nov13/cysticBN.png" width="500" title="Template for Genetic Bayesian Network" /></p>

<div class="bogus-wrapper"><notextile><figure class="code"><figcaption><span>creating genetic Bayesian network</span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class="line-number">1</span>
<span class="line-number">2</span>
<span class="line-number">3</span>
<span class="line-number">4</span>
<span class="line-number">5</span>
<span class="line-number">6</span>
<span class="line-number">7</span>
<span class="line-number">8</span>
<span class="line-number">9</span>
<span class="line-number">10</span>
<span class="line-number">11</span>
<span class="line-number">12</span>
<span class="line-number">13</span>
<span class="line-number">14</span>
<span class="line-number">15</span>
<span class="line-number">16</span>
<span class="line-number">17</span>
<span class="line-number">18</span>
<span class="line-number">19</span>
<span class="line-number">20</span>
</pre></td><td class="code"><pre><code class="python"><span class="line"><span class="n">family_tree</span> <span class="o">=</span> <span class="p">{}</span>
</span><span class="line"><span class="n">family_tree</span><span class="p">[</span><span class="s">&#39;Ira&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="bp">None</span><span class="p">,</span> <span class="bp">None</span><span class="p">]</span>
</span><span class="line"><span class="n">family_tree</span><span class="p">[</span><span class="s">&#39;Robin&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="bp">None</span><span class="p">,</span> <span class="bp">None</span><span class="p">]</span>
</span><span class="line"><span class="n">family_tree</span><span class="p">[</span><span class="s">&#39;Aaron&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="bp">None</span><span class="p">,</span> <span class="bp">None</span><span class="p">]</span>
</span><span class="line"><span class="n">family_tree</span><span class="p">[</span><span class="s">&#39;Rene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="bp">None</span><span class="p">,</span> <span class="bp">None</span><span class="p">]</span>
</span><span class="line"><span class="n">family_tree</span><span class="p">[</span><span class="s">&#39;James&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;Ira&#39;</span><span class="p">,</span> <span class="s">&#39;Robin&#39;</span><span class="p">]</span>
</span><span class="line"><span class="n">family_tree</span><span class="p">[</span><span class="s">&#39;Eva&#39;</span><span class="p">]</span>   <span class="o">=</span> <span class="p">[</span><span class="s">&#39;Ira&#39;</span><span class="p">,</span> <span class="s">&#39;Robin&#39;</span><span class="p">]</span>
</span><span class="line"><span class="n">family_tree</span><span class="p">[</span><span class="s">&#39;Sandra&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;Aaron&#39;</span><span class="p">,</span> <span class="s">&#39;Eva&#39;</span><span class="p">]</span>
</span><span class="line"><span class="n">family_tree</span><span class="p">[</span><span class="s">&#39;Jason&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;James&#39;</span><span class="p">,</span> <span class="s">&#39;Rene&#39;</span><span class="p">]</span>
</span><span class="line"><span class="n">family_tree</span><span class="p">[</span><span class="s">&#39;Benito&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;James&#39;</span><span class="p">,</span> <span class="s">&#39;Rene&#39;</span><span class="p">]</span>
</span><span class="line">
</span><span class="line"><span class="n">frequency_of_alleles_in_general_population</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">]</span>
</span><span class="line"><span class="n">probability_of_trait_based_on_genotype</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">]</span>
</span><span class="line">
</span><span class="line"><span class="n">TRAIT_PRESENT</span><span class="p">,</span> <span class="n">TRAIT_ABSENT</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">]</span>
</span><span class="line"><span class="n">FF</span><span class="p">,</span> <span class="n">Ff</span><span class="p">,</span> <span class="n">ff</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">]</span>
</span><span class="line">
</span><span class="line"><span class="c"># building the network</span>
</span><span class="line">
</span><span class="line"><span class="n">GN</span> <span class="o">=</span> <span class="n">geneticNetwork</span><span class="p">(</span><span class="n">family_tree</span><span class="p">,</span> <span class="n">frequency_of_alleles_in_general_population</span><span class="p">,</span> <span class="n">probability_of_trait_based_on_genotype</span><span class="p">)</span>
</span></code></pre></td></tr></table></div></figure></notextile></div>

<p>The main problem is that, so far, the way we are computing probabilities is by multiplying factors together, which means creating a CPD for the whole universe of possible assignments of all random variables (nodes). This super table takes into consideration all possible states, all possible scenarios, and therefore is huge. It used to take a very long time to compute it when using loops for the factor operations (a matter of hours), now with a vectorized version (see later post) it takes less than 3 seconds.</p>

<div class="bogus-wrapper"><notextile><figure class="code"><figcaption><span>running the network</span></figcaption><div class="highlight"><table><tr><td class="gutter"><pre class="line-numbers"><span class="line-number">1</span>
<span class="line-number">2</span>
<span class="line-number">3</span>
<span class="line-number">4</span>
<span class="line-number">5</span>
<span class="line-number">6</span>
<span class="line-number">7</span>
<span class="line-number">8</span>
<span class="line-number">9</span>
<span class="line-number">10</span>
<span class="line-number">11</span>
<span class="line-number">12</span>
<span class="line-number">13</span>
<span class="line-number">14</span>
<span class="line-number">15</span>
<span class="line-number">16</span>
</pre></td><td class="code"><pre><code class="python"><span class="line"><span class="c"># Evidence conditioning</span>
</span><span class="line"><span class="n">modify_Factor_by_evidence</span><span class="p">(</span><span class="s">&#39;Ira&#39;</span><span class="p">,</span>   <span class="s">&#39;pheno&#39;</span><span class="p">,</span> <span class="n">TRAIT_PRESENT</span><span class="p">)</span>
</span><span class="line"><span class="n">modify_Factor_by_evidence</span><span class="p">(</span><span class="s">&#39;James&#39;</span><span class="p">,</span> <span class="s">&#39;geno&#39;</span><span class="p">,</span> <span class="n">Ff</span><span class="p">)</span>
</span><span class="line"><span class="n">modify_Factor_by_evidence</span><span class="p">(</span><span class="s">&#39;Rene&#39;</span><span class="p">,</span>  <span class="s">&#39;geno&#39;</span><span class="p">,</span> <span class="n">FF</span><span class="p">)</span>
</span><span class="line">
</span><span class="line"><span class="c"># lets try first the whole kahuna CPD and compute the prob of developing CF</span>
</span><span class="line"><span class="n">a</span> <span class="o">=</span> <span class="n">build_joint_cpd</span><span class="p">()</span>
</span><span class="line">
</span><span class="line"><span class="c"># marginalizing</span>
</span><span class="line"><span class="n">target</span> <span class="o">=</span> <span class="n">GN</span><span class="p">[</span><span class="s">&#39;Benito&#39;</span><span class="p">][</span><span class="s">&#39;var_pheno&#39;</span><span class="p">]</span>
</span><span class="line"><span class="n">lista</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">a</span><span class="o">.</span><span class="n">variables</span> <span class="k">if</span> <span class="n">x</span><span class="o">!=</span><span class="n">target</span><span class="p">]</span>
</span><span class="line"><span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">lista</span><span class="p">:</span>
</span><span class="line">    <span class="n">a</span> <span class="o">=</span> <span class="n">FactorOperations</span><span class="o">.</span><span class="n">marginalize</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span>
</span><span class="line">
</span><span class="line"><span class="c"># the result</span>
</span><span class="line"><span class="k">print</span> <span class="s">&#39;probability of Benito showing ailment&#39;</span><span class="p">,</span> <span class="mf">100.</span><span class="o">*</span><span class="n">a</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s">&#39;%&#39;</span>
</span></code></pre></td></tr></table></div></figure></notextile></div>

<p>In an initial state, without more information, the probability for Benito developing Cystic Fibrosis is 19.70%. It is not even necessary to infer this from the network (although it is possible) since it is the weighted average of the genotype distribution <code>0.197 = (0.01*0.8)+(0.18*0.6)+(0.81*0.1)</code>.</p>

<p>Now, if his grandfather Ira had the ailment, (the trait was present as expressed in phenotype and suffered the disease), the chances of Benito being in the same situation raises to 24.75%</p>

<p>Furthermore, his parents did a genetic test prior to having children and they know that Jame’s genotype is Ff and that Rene’s is FF. Now the probability has risen again to 70%.</p>

</div>


  <footer>
    <p class="meta">
      
  

<span class="byline author vcard">Posted by <span class="fn">Javier Soto</span></span>

      








  


<time datetime="2013-11-04T08:01:00-08:00" pubdate data-updated="true">Nov 4<span>th</span>, 2013</time>
      

<span class="categories">
  
    <a class='category' href='/blog/categories/pgn/'>PGN,</a>, <a class='category' href='/blog/categories/genetic/'>genetic</a>
  
</span>


    </p>
    
      <div class="sharing">
  
  <a href="http://twitter.com/share" class="twitter-share-button" data-url="http://sotoseattle.github.io/blog/2013/11/04/GBN2/" data-via="" data-counturl="http://sotoseattle.github.io/blog/2013/11/04/GBN2/" >Tweet</a>
  
  
  
</div>

    
    <p class="meta">
      
        <a class="basic-alignment left" href="/blog/2013/11/03/Genetic-BN/" title="Previous Post: Genetic Bayesian Network (I)">&laquo; Genetic Bayesian Network (I)</a>
      
      
        <a class="basic-alignment right" href="/blog/2013/11/05/GBN3/" title="Next Post: Genetic Bayesian Network (III)">Genetic Bayesian Network (III) &raquo;</a>
      
    </p>
  </footer>
</article>

  <section>
    <h1>Comments</h1>
    <div id="disqus_thread" aria-live="polite"><noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
</div>
  </section>

</div>

<aside class="sidebar">
  
    <section>
  <h1>Recent Posts</h1>
  <ul id="recent_posts">
    
      <li class="post">
        <a href="/blog/2015/12/05/ruby-continuations/">Ruby Continuations</a>
      </li>
    
      <li class="post">
        <a href="/blog/2015/03/04/Seattlerb-GoL/">Seattle Ruby Game Of Life</a>
      </li>
    
      <li class="post">
        <a href="/blog/2015/03/01/CodeFellows/">Article Published</a>
      </li>
    
      <li class="post">
        <a href="/blog/2015/02/15/Scheme-GOL/">Game of Life in Scheme</a>
      </li>
    
      <li class="post">
        <a href="/blog/2014/11/18/4-PSD/">4 Principles of Simple Design Reflections</a>
      </li>
    
  </ul>
</section>





  
</aside>


    </div>
  </div>
  <footer role="contentinfo"><p>
  Copyright &copy; 2015 -  Javier Soto <br/>
  <span class="credit">Powered by <a href="http://octopress.org">Octopress</a> + <a href="https://github.com/ioveracker/mnml">mnml</a>.
	  
  </span>
</p>

</footer>
  

<script type="text/javascript">
      var disqus_shortname = 'sotoseattle';
      
        
        // var disqus_developer = 1;
        var disqus_identifier = 'http://sotoseattle.github.io/blog/2013/11/04/GBN2/';
        var disqus_url = 'http://sotoseattle.github.io/blog/2013/11/04/GBN2/';
        var disqus_script = 'embed.js';
      
    (function () {
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = 'http://' + disqus_shortname + '.disqus.com/' + disqus_script;
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    }());
</script>







  <script type="text/javascript">
    (function(){
      var twitterWidgets = document.createElement('script');
      twitterWidgets.type = 'text/javascript';
      twitterWidgets.async = true;
      twitterWidgets.src = 'http://platform.twitter.com/widgets.js';
      document.getElementsByTagName('head')[0].appendChild(twitterWidgets);
    })();
  </script>





</body>
</html>
<!-- mathjax config similar to math.stackexchange -->
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  jax: ["input/TeX", "output/HTML-CSS"],
  tex2jax: {
    inlineMath: [ ['$', '$'] ],
    displayMath: [ ['$$', '$$']],
    processEscapes: true,
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
  },
  messageStyle: "none",
  "HTML-CSS": { preferredFont: "TeX", availableFonts: ["STIX","TeX"] }
});
</script>
<script src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML" type="text/javascript"></script>
